File: ios/App/Sources/Services/Transcriber.swift
import Foundation

struct ServiceError: LocalizedError {
    let message: String
    var errorDescription: String? { message }
}

/// Транскрипция през OpenAI-съвместим endpoint `POST {base}/v1/audio/transcriptions`.
/// Работи с whisper.cpp сървър, faster-whisper-server, OpenAI и подобни.
struct Transcriber {
    let settings: Settings

    func transcribe(fileURL: URL) async throws -> String {
        let base = settings.asrBaseURL.trimmingCharacters(in: .whitespaces)
            .trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        guard !base.isEmpty, let url = URL(string: "\(base)/v1/audio/transcriptions") else {
            throw ServiceError(message: String(localized: "No transcription server configured (Settings → Transcription)."))
        }
        let audio = try Data(contentsOf: fileURL)
        guard !audio.isEmpty else {
            throw ServiceError(message: String(localized: "The audio file is empty."))
        }

        let boundary = "callscribe-\(UUID().uuidString)"
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 180
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        if !settings.asrAPIKey.isEmpty {
            request.setValue("Bearer \(settings.asrAPIKey)", forHTTPHeaderField: "Authorization")
        }

        var body = Data()
        func appendField(_ name: String, _ value: String) {
            body.append("--\(boundary)\r\n".data(using: .utf8)!)
            body.append("Content-Disposition: form-data; name=\"\(name)\"\r\n\r\n".data(using: .utf8)!)
            body.append("\(value)\r\n".data(using: .utf8)!)
        }

        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append(
            "Content-Disposition: form-data; name=\"file\"; filename=\"\(fileURL.lastPathComponent)\"\r\n"
                .data(using: .utf8)!
        )
        body.append("Content-Type: audio/m4a\r\n\r\n".data(using: .utf8)!)
        body.append(audio)
        body.append("\r\n".data(using: .utf8)!)

        appendField("model", settings.asrModel.isEmpty ? "whisper-1" : settings.asrModel)
        appendField("language", settings.language.isEmpty ? "bg" : settings.language)
        appendField("response_format", "json")
        body.append("--\(boundary)--\r\n".data(using: .utf8)!)
        request.httpBody = body

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw ServiceError(message: String(localized: "No response from the transcription server."))
        }
        guard (200..<300).contains(http.statusCode) else {
            let detail = String(data: data, encoding: .utf8)?.prefix(300) ?? ""
            throw ServiceError(message: String(format: String(localized: "Transcription failed (HTTP %d): %@"), http.statusCode, String(detail)))
        }

        if let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let text = object["text"] as? String {
            return text.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        return (String(data: data, encoding: .utf8) ?? "")
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }
}

File: app/src/main/java/com/callscribe/ai/Transcriber.kt
package com.callscribe.ai

import android.content.Context
import com.callscribe.R
import com.callscribe.data.Settings
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.Request
import okhttp3.RequestBody.Companion.asRequestBody
import org.json.JSONObject
import java.io.File

/**
 * Транскрипция през OpenAI-съвместим endpoint `POST {base}/v1/audio/transcriptions`.
 * Работи с локален whisper.cpp сървър, faster-whisper-server, OpenAI и т.н.
 */
class Transcriber(private val context: Context, private val settings: Settings) {

    fun transcribe(file: File): String {
        val base = settings.asrBaseUrl
        if (base.isBlank()) throw AiException(context.getString(R.string.err_no_asr))
        if (!file.exists() || file.length() == 0L) throw AiException(context.getString(R.string.err_empty_audio))

        val body = MultipartBody.Builder()
            .setType(MultipartBody.FORM)
            .addFormDataPart("file", file.name, file.asRequestBody("audio/mp4".toMediaType()))
            .addFormDataPart("model", settings.asrModel.ifBlank { "whisper-1" })
            .addFormDataPart("language", settings.language.ifBlank { "bg" })
            .addFormDataPart("response_format", "json")
            .build()

        val builder = Request.Builder()
            .url("$base/v1/audio/transcriptions")
            .post(body)
        if (settings.asrApiKey.isNotBlank()) {
            builder.header("Authorization", "Bearer ${settings.asrApiKey}")
        }

        Http.client.newCall(builder.build()).execute().use { response ->
            val text = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                throw AiException(context.getString(R.string.err_transcribe, response.code, text.take(300)))
            }
            return try {
                JSONObject(text).optString("text").trim()
            } catch (e: Exception) {
                text.trim()
            }
        }
    }
}
